test/controllers/tasks_controller_test.rb

```diff
@@ -71,56 +64,126 @@
           completion_date: nil,
         },
       }
-
+      
       # Act-Assert
       expect {
         post tasks_path, params: task_hash
       }.must_change "Task.count", 1
-
+      
       new_task = Task.find_by(name: task_hash[:task][:name])
       expect(new_task.description).must_equal task_hash[:task][:description]
-      expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
-      expect(new_task.completed).must_equal task_hash[:task][:completed]
-
+      # expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
+      # expect(new_task.completion_date).must_equal task_hash[:task][:completion_date]
+      expect(new_task.completion_date).must_be_nil task_hash[:task][:completion_date]
+      
       must_respond_with :redirect
-      must_redirect_to task_path(new_task.id)
+      must_redirect_to root_path
     end
   end
-
+  
   # Unskip and complete these tests for Wave 3
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      
+      get edit_task_path(task.id)
+      
+      must_respond_with :success
     end
-
+    
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      # invalid_id = task_path(-1)
+      
+      get edit_task_path(8923749234)
+      
+      must_respond_with :redirect
+      must_redirect_to root_path
     end
   end
-
+  
   # Uncomment and complete these tests for Wave 3
   describe "update" do
+    before do
+      @new_task = Task.create(name: "Laundry", description: "Wash and dry", completion_date: nil)
+    end
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
+    updated_task_form_data = {
+      task: {
+        name: "Clean Windows",
+        description: 'Use windex',
+        completion_date: Time.now + 7.days
+      }
+    }
     it "can update an existing task" do
-      # Your code here
+      expect {
+        patch task_path(@new_task.id), params: updated_task_form_data
+      }.wont_change "Task.count"
+      
+      @new_task.reload
+      
+      expect(@new_task.name).must_equal "Laundry"
     end
-
+    
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      
+      patch task_path(2348734)
+      must_respond_with :redirect
+      must_redirect_to root_path
     end
   end
-
+  
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
-
-  end
-
-  # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
-  end
-end
+    it "can destroy a task" do
```

Comment:
You need to test `destroy` and `toggle_complete` and `toggle_incomplete` for valid and **invalid** task ids.